irods/client_configuration/__init__.py

```diff
@@ -0,0 +1,223 @@
+    a configuration file.
+
+    The 'root' and 'string' parameters are not likely to be overridden when called from an
+    application. They should usually only vary from the defaults when save() recurses into itself.
+    However, for due explanation's sake: 'root' specifies at which subtree node to start writing,
+    None denoting the top level; and 'string' specifies a prefix for the dotted prefix name,
+    which should be empty for an invocation that references the settings' top level namespace.
+    Both of these defaults are in effect when calling save() without explicit parameters.
+
+    The configuration file path will normally be the value of DEFAULT_CONFIG_PATH,
+    but this can be overridden by supplying a nonzero string in the 'file' parameter.
+    """
+    _file = None
+    auto_close_settings = False
+    try:
+        if not file:
+            from .. import get_settings_path
+            file = get_settings_path()
+        if isinstance(file,str):
+            _file = open(file,'w')
+            auto_close_settings = True
+        else:
+            _file = file # assume file-like object if not a string
+        if root is None:
+            root = sys.modules[__name__]
+        for k,v in _var_items(root):
+            dotted_string = string + ("." if string else "") + k
+            if isinstance(v,iRODSConfiguration):
+                save(root = v, string = dotted_string, file = _file)
+            else:
+                print(dotted_string, repr(v), sep='\t\t', file = _file)
+        return file
+    finally:
+        if _file and auto_close_settings:
+            _file.close()
+
+def _load_config_line(root, setting, value):
+    arr = [_.strip() for _ in setting.split('.')]
+    # Compute the object referred to by the dotted name.
+    attr = ''
+    for i in filter(None,arr):
+        if attr:
+            root = getattr(root,attr)
+        attr = i
+    # Assign into the current setting of the dotted name (effectively <root>.<attr>)
+    # using the loaded value.
+    if attr:
+        return setattr(root, attr, ast.literal_eval(value))
+    error_message = 'Bad setting: root = {root!r}, setting = {setting!r}, value = {value!r}'.format(**locals())
+    raise RuntimeError (error_message)
+
+# The following regular expression is used to match a configuration file line of the form:
+# ---------------------------------------------------------------
+#         <optional whitespace>
+#  key:   <dotted-name specification>
+#         <whitespace of length 1 or more>
+#  value: <A Python value which can be given to ast.literal_eval(); e.g. 5, True, or 'some_string'>
+#         <optional whitespace>
+
+_key_value_pattern = re.compile(r'\s*(?P<key>\w+(\.\w+)+)\s+(?P<value>\S.*?)\s*$')
+
+class _ConfigLoadError:
+    """
+    Exceptions that subclass this type can be thrown by the load() function if
+    their classes are listed in the failure_modes parameter of that function.
+    """
```

Comment:
Is this syntactically valid? I noticed that there used to be a `pass` here...

Reply:
It is syntactically valid because of the docstring.  believe it or not. 
Not only that, codacy doesn't want to see the "pass" there anymore... so I removed it.
One thing to realize, too, is that the docstring is valid code, and its value gets inserted into the target class's / function's __doc__ attribute at runtime.